✨(backend) add API enpoints for pairing external devices to live videos

kernicPanel opened the pull request:
## Purpose

We want to allow pairing external devices to Marsha live. External devices are cameras or set-top boxes that can act as video and sound sources for a live in Marsha.

## Proposal
Initialy in #1196 we thought storing secrets in live_info video field.
Instead a new model has been added to store them for performance purpose during challenge.

- [x] Add an API endpoint at `/api/video/<id>/pairing-secret/` which
  - generates a unique secret
  - stores it in the LivePairing model with an `expires_at` timestamp
  - responds with a json containing the generated secret
- [x] Add an API endpoint at `/api/video/pairing-challenge` which
  - receives the secret, and the external device id
  - checks if the secret exists and if it is not expired, and throttles requests to avoid brute force attacks,
  - responds with a json containing the jitsi url of the matching video live

Fixes #1196

sampaccoud commented on src/backend/marsha/core/models/video.py:
```diff
@@ -469,3 +471,84 @@ class Meta:
+
+    def generate_secret(self):
+        """Stores generated secret and expiration date."""
+        existing_secrets = LivePairing.objects.values_list("secret", flat=True)
```
I would say this is overkill and can cause race conditions. A try/except loop may be appropriate here?
```
for _ in range(2):
    self.secret = secrets.randint(1000000)
    try:
       self.save()
    except ...:
        pass
    else:
        break
```
If we fall on an existing secret twice in a row, it is a sign that we have too many challenges ongoing? Do you think we need to formally limit the number of challenges being run in parallel in order to ensure a very low probability of finding a secret by chance?

kernicPanel replied:
Sorry, but I don't get which race condition would be caused, and in which way the approach I propose is overkill…
Can you elaborate on this, please ?

sampaccoud replied:
Because two concurrent requests could generate a new code that would be the same... So the second one would fail upon db transaction commit. We also need to detect if too may pairing challenges are open simultaneously as it would kill the security of the process?

kernicPanel replied:
Regarding the current code state, and discussions we had, is your comment still relevant ?

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +475,105 @@ def update_live_state(
+        return Response({"detail": "Secret expired."}, status=410)
+
+    try:
+        jitsi_domain = live_pairing.video.live_info.get("jitsi").get("domain")
```
This is wrong. The `jitsi` domain is never saved in the `video` model. It is computed in the `serializer` https://github.com/openfun/marsha/blob/69431e5ae0663c1a779556a7575352d803688d26/src/backend/marsha/core/serializers/video.py#L669

You must use `settings.JITSI_DOMAIN` here.

lunika commented on src/backend/marsha/core/migrations/0026_auto_20211117_1528.py:
```diff
@@ -0,0 +1,789 @@
+import django.utils.timezone
+
+
+class Migration(migrations.Migration):
```
Only keep the new model creation in the migration and remove everything else

lunika commented on src/backend/marsha/settings.py:
```diff
@@ -158,6 +158,9 @@ class Base(Configuration):
         "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.LimitOffsetPagination",
         "PAGE_SIZE": 50,
+        "DEFAULT_THROTTLE_RATES": {
+            "anon": "2/minute",
```
`2` it's a little bit harsh @sampaccoud , you have most of the time 3 attempts to use a valid PIN, we are in a similar case IMO

sampaccoud replied:
I'm ok with 3 of course :+1: For a known device that is.

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +475,104 @@ def update_live_state(
+        """
+        video = self.get_object()
+        try:
+            video.live_info.get("jitsi").get("domain")
```
No it's wrong. I added a comment in my review explaining how to have this value.

kernicPanel commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +475,104 @@ def update_live_state(
+
+    video_id = live_pairing.video.id
+    live_pairing.delete()
+    Device.objects.create(id=request.data.get("box_id"))
```
Would `get_or_create` be ok, as we don't need to get it ?

sampaccoud replied:
I would say yes... or a try/except? And add a test for this situation :wink:

kernicPanel commented on src/backend/marsha/core/migrations/0027_device.py:
```diff
@@ -0,0 +1,55 @@
+# Generated by Django 3.2.9 on 2021-11-23 14:11
```
As you want, but the idea here was to separate commits nicely.
But I can group them in the first one, just let me know.

sampaccoud replied:
Looking at the 2 commits I think there should be only one commit since it is the same feature. One e 
Ndpoint does not make sense without the other one.

kernicPanel commented on src/backend/marsha/core/models/video.py:
```diff
@@ -469,3 +471,84 @@ class Meta:
         verbose_name = _("live registration")
+
+
+class LivePairing(BaseModel):
```
I didn't think of it…
Adding `_safedelete_policy = HARD_DELETE_NOCASCADE` to this model would be ok, WDYT ?

kernicPanel commented:
@sampaccoud @lunika I pushed fixups that addresses your comments, except for the migration.
I think it will be easier for you to review changes without rebasing.
I will commit only one migration after your new review.

carofun commented on src/backend/marsha/core/api/video.py:
```diff
@@ -1,18 +1,30 @@
 """Declare API endpoints for videos with Django RestFramework viewsets."""
```
in the commit message
add an API en**d**point for pairing secret request"

carofun commented on src/backend/marsha/core/api/video.py:
```diff
@@ -12,7 +12,14 @@
+    Thumbnail,
+    TimedTextTrack,
+    Video,
+    LivePairing,
```
please order in alphabetical order

lunika commented on src/backend/marsha/core/tests/test_api_video_live_pairing.py:
```diff
@@ -0,0 +1,354 @@
+
+    def test_api_video_pairing_challenge_valid_secret_non_jitsi(self):
+        """Request with secret matching a non jitsi video.
+        Should delete live pairing and raise a 410."""
```
```suggestion
        Should delete live pairing and raise a 400."""
```

lunika commented on src/backend/marsha/core/tests/test_api_video_live_pairing.py:
```diff
@@ -0,0 +1,354 @@
+        self.assertEqual(device.id, box_id)
+
+    def test_api_video_pairing_challenge_unknown_devices_throttling(self):
+        """Throttling should prevent more than two requests per minute.
```
```suggestion
        """Throttling should prevent more than three requests per minute.
```

lunika commented on src/backend/marsha/core/tests/test_api_video_live_pairing.py:
```diff
@@ -0,0 +1,354 @@
+            {"detail": "Request was throttled. Expected available in 60 seconds."},
+        )
+
+        # resetting throttling by mocking timer used by DRF in AnonRateThrottle
```
calling `cache.clear()` is not easier ?

kernicPanel replied:
Yes, it would be easier, but i think it's more realistic like this.

kernicPanel replied:
Are you ok with this ?

sampaccoud commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +477,102 @@ def update_live_state(
+    if live_pairing.video.live_type != JITSI:
+        return Response({"detail": "Matching video is not a Jitsi Live."}, status=400)
+
+    jitsi_domain = settings.JITSI_DOMAIN
```
I would use SETTINGS.JITSI_DOMAIN directly in the f-string

sampaccoud commented on src/backend/marsha/settings.py:
```diff
@@ -158,6 +158,9 @@ class Base(Configuration):
         "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.LimitOffsetPagination",
         "PAGE_SIZE": 50,
+        "DEFAULT_THROTTLE_RATES": {
+            "anon": "3/minute",
```
So 3/minutes during 3 minutes is not 3 attempts like we discussed earlier but 9 attempts? In this case it should be set to 1/minute or expiry should be set to 1 minute?

kernicPanel replied:
We still need to discuss this point.
Maybe looking at related tests would help us to find a good balance ?

sampaccoud commented on src/backend/marsha/urls.py:
```diff
@@ -62,6 +63,7 @@
     path("videos/<uuid:uuid>", VideoView.as_view(), name="video_public"),
     path("documents/<uuid:uuid>", DocumentView.as_view(), name="document_public"),
     # API
+    path("api/pairing-challenge", pairing_challenge, name="pairing_challenge"),
```
How come there's no url for pairing request?

kernicPanel replied:
As a secret request is related to a video, the endpoint is `/api/videos/{video.id}/pairing-secret/`

sampaccoud replied:
True

sampaccoud commented on src/backend/marsha/core/tests/test_models_live_pairing.py:
```diff
@@ -0,0 +1,35 @@
+        # Remove one from the list
+        existing_secrets.remove("123456")
+
+        # Ensure removed secret has been found and used
```
Actually this test is scary because it demonstrates a case where the secret is directly guessable 😅 as mentioned above we need to protect against this situation.

kernicPanel replied:
I don't get how the secret is guessable…

sampaccoud replied:
Because if all possible secrets are existing, then whatever secret I submit, it will match some video... In order for this pairing to be secure, we must limit the number of existing secrets at a given time... maybe 10 or 100? And you shouldn't return a different error if the secret has expired or is unknown or has throttled... always return a 404 is safer.

sampaccoud replied:
I thing the try/except I proposed instead of looking for the existing secrets, is solving this issue because the try/except will fail when the number of secrets is too much...

lunika replied:
What about creating a management command deleting all secret expired every hour ? Using redis we can use an expire key 😁

kernicPanel replied:
I don't think `secret_generator` should be responsible for security aspects. His job is only to return an available secret.
And, I agree, if 999 999 secrets are existing, we have a problem which should be addressed, but not here.
This test reflects only the performance of my implementation.

sampaccoud replied:
Expired secrets do not pause any problem if you don't leak the information on the error and if you add a conditional unique constraint that only checks unicity on valid secrets?

kernicPanel replied:
secrets are already unique, see LivePairing model :
```
    secret = models.CharField(
        max_length=6,
        unique=True,
        verbose_name=_("secret"),
        help_text=_("live pairing secret"),
    )
```
Maybe we could  simply delete all expired secrets on every request ? I think it should be fast enough, and not noticed by the users.
But I agree with you, we shouldn't give any clues on errors.

sampaccoud replied:
Yes good idea, start by deleting all expired secrets and let's avoid another cron :sweat_smile:

kernicPanel replied:
Are we ok with the code related to this thread ?

sampaccoud replied:
OK for the delete but I still think the code retrieving existing secrets does not work with concurrent requests and should be based on a try/except pattern.

kernicPanel replied:
Should we test concurrent requests then ?

sampaccoud replied:
well... that's hard... but the try/except pattern surely solves it.

sampaccoud commented on src/backend/marsha/core/tests/test_models_live_pairing.py:
```diff
@@ -0,0 +1,35 @@
+
+class LivePairingModelsTestCase(TestCase):
+    """Test our intentions about the LivePairing model."""
+
```
I think you should test the uniqueness of the secret field

sampaccoud commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +477,102 @@ def update_live_state(
+        return Response(serializer.errors, status=400)
+
+    try:
+        live_pairing = LivePairing.objects.get(secret=request.data.get("secret"))
```
If I'm not mistaken you forgot to use the box I'd so you are not blocking unknown boxes from making many requests...

kernicPanel replied:
Thanks to @lunika , I added a test which should cover this case : `test_api_video_pairing_challenge_unknown_device_throttling` WDYT ?

kernicPanel replied:
Looking at tests, are we ok with this ?

sampaccoud commented on src/backend/marsha/core/tests/test_api_video_live_pairing.py:
```diff
@@ -0,0 +1,401 @@
+        )
+
+        # other request should be throttled
+        response = self.client.post(
```
Can  you explain why it was throttled since you configured the throttle to check the box ID? My understanding is that a new box id should be throttled independently of the first one.

kernicPanel replied:
This was indeed the case, but I changed this recently :
If the box is known, the throttling uses the box id as a key.
Otherwise, no key is set, so every unknown request shares the same throttling rate.

sampaccoud commented on src/backend/marsha/core/models/video.py:
```diff
@@ -467,3 +471,82 @@ class Meta:
+        return self.video.title
+
+    @classmethod
+    def delete_expired(cls):
```
Actually, this should be on the manager as a new query

sampaccoud commented on src/backend/marsha/core/serializers/video.py:
```diff
@@ -288,6 +295,22 @@ class InitLiveStateSerializer(serializers.Serializer):
+    class Meta:  # noqa
+        model = LivePairing
+        fields = ("secret",)
+        read_only_fields = ("secret",)
```
You should add a test trying to patch or put the object modifyng the secret field

kernicPanel commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +478,110 @@ def update_live_state(
+            live_pairing = LivePairing.objects.get(video=video)
+        except LivePairing.DoesNotExist:
+            live_pairing = LivePairing(video=video)
+        live_pairing.generate_secret()
+        live_pairing.save()
```
```suggestion
        for _ in range(2):
            try:
                live_pairing.generate_secret()
                live_pairing.save()
                break
            except ValidationError:
                pass
```
Let's try a compromise @sampaccoud :
Here we're keeping secret generation as is, but we're adding a second attempt to save it at the API level.
WDYT ?

sampaccoud replied:
ok I will approve this version but I still find that the `generate_secret` method is complicated for nothing when a simple try/except does the job :wink:

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +478,116 @@ def update_live_state(
+            box_id = request.data.get("box_id")
+            Device.objects.get(pk=box_id)
+            return box_id
+        except (Device.DoesNotExist, ValidationError):
```
Why `ValidationError` ?

kernicPanel replied:
Because sending a wrong box_id (that is not an uuid), like in `test_api_video_pairing_challenge_wrong_box_id`, would raise it.

lunika replied:
:+1:

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +478,116 @@ def update_live_state(
+    except LivePairing.DoesNotExist:
+        return Response({"detail": "Secret not found."}, status=404)
+
+    if live_pairing.video.live_type != JITSI:
```
If it's not a `JITSI` video, the `live_pairing` should be deleted then ?

kernicPanel replied:
Indeed!

lunika replied:
So I think deleting the `live_pairing` should be moved just before this `if`. WDYT ?

kernicPanel replied:
That's how I managed this, yes.

lunika commented on src/backend/marsha/settings.py:
```diff
@@ -296,6 +299,7 @@ class Base(Configuration):
     JITSI_DOMAIN = values.Value("meet.jit.si")
     JITSI_CONFIG_OVERWRITE = values.DictValue({})
     JITSI_INTERFACE_CONFIG_OVERWRITE = values.DictValue({})
+    JITSI_SECRET_EXPIRATION = 3
```
This is not directly related to jitsi. IMO you should not orefix it with `JITSI`. This variable name hould reflect the `live_pairing` feature. `LIVE_SECRET_EXPIRATION_MINUTES` ? `LIVE_SECRET_EXPIRATION_TIME` ? `LIVE_SECRET_EXPIRATION_DURATION` ?

kernicPanel replied:
I renamed it to LIVE_PAIRING_EXPIRATION_MINUTES

sampaccoud replied:
For coherence with other settings, we could count in seconds wdyt?

kernicPanel commented on src/backend/marsha/core/api/video.py:
```diff
@@ -465,6 +478,110 @@ def update_live_state(
+            box_id = request.data.get("box_id")
+            Device.objects.get(pk=box_id)
+            return box_id
+        except (Device.DoesNotExist, ValidationError):
```
It surely do : if I remove it this happens :
```
===================================================== FAILURES =====================================================
________________________ PairingDeviceAPITest.test_api_video_pairing_challenge_wrong_box_id ________________________

self = <marsha.core.tests.test_api_video_live_pairing.PairingDeviceAPITest testMethod=test_api_video_pairing_challenge_wrong_box_id>

    def test_api_video_pairing_challenge_wrong_box_id(self):
        """Request with wrong box_id payload should raise a 400 error."""
        payload = {"box_id": "wrong", "secret": "000000"}
        response = self.client.post("/api/pairing-challenge", data=payload)
    
        self.assertEqual(response.status_code, 400)
>       self.assertEqual(response.json(), {"box_id": ["Must be a valid UUID."]})
E       AssertionError: ['“%(value)s” is not a valid UUID.'] != {'box_id': ['Must be a valid UUID.']}

marsha/core/tests/test_api_video_live_pairing.py:244: AssertionError
```

sampaccoud replied:
Ok then it should be handled in the serializer in a "clean_box_id" method.

sampaccoud replied:
https://www.django-rest-framework.org/api-guide/serializers/#field-level-validation

kernicPanel replied:
Did you mean validate_box_id ? If it is, isn't it already validated ?

sampaccoud replied:
It should be but you must instanciate PairingChallengeSerializer and validate it instead of poking request.data.get("box_id")

sampaccoud reviewed (approved):
:clap: :tada:
